contributors/design-proposals/storage/container-storage-interface.md

```diff
@@ -0,0 +1,458 @@
+
+The Kubernetes volume sub-system does not currently support block volumes (only file), so for alpha, the Kubernetes CSI volume plugin will only support file.
+
+#### Attaching and Detaching
+
+The attach/detach controller,running as part of the kube-controller-manager binary on the master, decides when a CSI volume must be attached or detached from a particular node.
+
+When the controller decides to attach a CSI volume, it will call the in-tree CSI volume plugin’s attach method. The in-tree CSI volume plugin’s attach method will do the following:
+
+1. Create a new `VolumeAttachment` object (defined in the “Communication Channels” section) to attach the volume.
+    * The name of the of the `VolumeAttachment` object will be the `pv-<PVName>-<NodeName>`.
+      * `pv-` prefix is used to allow using other scheme(s) for inline volumes in future, with their own prefix.
+      * `PVName` is `PV.name` of the attached PersistentVolume.
+      * `NodeName` is `Node.name` of the node where the volume should be attached to.
+    * If a `VolumeAttachment` object with the corresponding name already exists, the in-tree volume plugin will simply begin to poll it as defined below. The object is not modified; only the external-attacher should change the status fields; and the external-attacher is responsible for it’s own retry and error handling logic.
+2. Poll the `VolumeAttachment` object waiting for one of the following conditions:
+    * The `VolumeAttachment.Status.Attached` field to become `true`.
+      * The operation completes successfully.
+    * An error to be set in the `VolumeAttachment.Status.AttachError` field.
+      * The operation terminates with the specified error.
+    * The operation to timeout.
+      * The operation terminates with timeout error.
+    * The `VolumeAttachment.DeletionTimestamp` is set.
+      * The operation terminates with an error indicating a detach operation is in progress.
+      * The `VolumeAttachment.Status.Attached` value must not be trusted. The attach/detach controller has to wait until the object is deleted by the external-attacher before creating a new instance of the object.
+
+When the controller decides to detach a CSI volume, it will call the in-tree CSI volume plugin’s detach method. The in-tree CSI volume plugin’s detach method will do the following:
+
+1. Delete the corresponding `VolumeAttachment` object (defined in the “Communication Channels” section) to indicate the volume should be detached.
+2. Poll the `VolumeAttachment` object waiting for one of the following conditions:
```

Comment:
If object is actually delete (as indicated earlier)... polling it will not work since object will not be in etcd.

Reply:
The object should be created if it does not exist (step 1).